### PR TITLE
validation check input

### PR DESCRIPTION
**Pull request by qshuai**
PTAL

### CODE REVIEW

File: blockchain/Validation.go [hunk 1]
```diff
@@ -2805,7 +2827,50 @@ func IsCurrentForFeeEstimation() bool {
...
+	// to enforce that mempool doesn't change between when we check the view and
+	// when we actually call through to CheckInputs
+	mpool.Mtx.Lock()
+	defer mpool.Mtx.Unlock()
+
+	if tx.IsCoinBase() {
+		panic("critical error")
+	}
+	for _, txin := range tx.Ins {
+		coin := cache.AccessCoin(txin.PreviousOutPoint)
```
Comment (ludete) [inline blockchain/Validation.go]: the variable name, I think use view is more corret.

File: blockchain/Validation.go [hunk 2]
```diff
@@ -2805,7 +2827,50 @@ func IsCurrentForFeeEstimation() bool {
...
+
+		txFrom := mpool.Get(&txin.PreviousOutPoint.Hash)
+		if txFrom != nil {
+			if txFrom.TxHash() != txin.PreviousOutPoint.Hash {
+				panic("critical error")
+			}
+			if len(txFrom.Outs) <= int(txin.PreviousOutPoint.Index) {
+				panic("critical error")
+			}
+			if txFrom.Outs[txin.PreviousOutPoint.Index] != coin.TxOut {
```
Comment (ludete) [inline blockchain/Validation.go]: I think you can add IsEuqal function to compare the two TxOut .

File: blockchain/Validation.go [hunk 3]
```diff
@@ -2805,7 +2827,50 @@ func IsCurrentForFeeEstimation() bool {
...
+				panic("critical error")
+			}
+		} else {
+			coinFromDisk := gcoinsTip.AccessCoin(txin.PreviousOutPoint)
+			if coinFromDisk.IsSpent() {
+				panic("critical error ")
+			}
+			if coinFromDisk.TxOut.IsEqual(coin.TxOut) {
```
Comment (ludete) [inline blockchain/Validation.go]: error

File: blockchain/Validation.go [hunk 4]
```diff
@@ -2819,7 +2884,182 @@ func CheckInputsFromMempoolAndCache(tx *model.Tx, state *model.ValidationState,
...
-	sigCacheStore bool, scirptCacheStore bool, txData *model.PrecomputedTransactionData, checks []*ScriptCheck) bool {
+	sigCacheStore bool, scriptCacheStore bool, txData *model.PrecomputedTransactionData, checks []*ScriptCheck) bool {
+
+	if tx.IsCoinBase() {
+		panic("critical error")
+	}
+	if !CheckTxInputs(tx, state, cache, GetSpendHeight(cache)) {
+		return false
+	}
+
```
Comment (ludete) [inline blockchain/Validation.go]: loss code

File: blockchain/Validation.go [hunk 5]
```diff
@@ -2819,7 +2884,182 @@ func CheckInputsFromMempoolAndCache(tx *model.Tx, state *model.ValidationState,
...
+	// We only use the first 19 bytes of nonce to avoid a second SHA round -
+	// giving us 19 + 32 + 4 = 55 bytes (+ 8 + 1 = 64)
+	if 55-unsafe.Sizeof(flags)-32 < 128/8 {
+		// compile error
+		panic("Want at least 128 bits of nonce for script execution cache")
+	}
+
+	k0 := utils.GetRand(math.MaxUint64)
+	k1 := utils.GetRand(math.MaxUint64)
+	hasher := utils.NewSipHasher(k0, k1)
```
Comment (ludete) [inline blockchain/Validation.go]: the hash you use error.

File: blockchain/Validation.go [hunk 6]
```diff
@@ -2819,7 +2884,182 @@ func CheckInputsFromMempoolAndCache(tx *model.Tx, state *model.ValidationState,
...
+func CheckTxInputs(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsViewCache, spendHeight int) bool {
+	// This doesn't trigger the DoS code on purpose; if it did, it would make it
+	// easier for an attacker to attempt to split the network.
+	if !cache.HaveInputs(tx) {
+		return state.Invalid(false, 0, "", "Inputs unavailable")
+	}
+
+	valueIn := btcutil.Amount(0)
+	fees := btcutil.Amount(0)
+	length := len(tx.Ins)
```
Comment (ludete) [inline blockchain/Validation.go]: this line code can ignore.

File: blockchain/Validation.go (hunk 1 above)
Comment (qshuai) [reply blockchain/Validation.go]: ok

File: blockchain/Validation.go (hunk 2 above)
Comment (qshuai) [reply blockchain/Validation.go]: yes

File: blockchain/Validation.go (hunk 4 above)
Comment (qshuai) [reply blockchain/Validation.go]: need to discuss

File: blockchain/Validation.go (hunk 3 above)
Comment (qshuai) [reply blockchain/Validation.go]: ?

File: blockchain/Validation.go (hunk 5 above)
Comment (qshuai) [reply blockchain/Validation.go]: yes, you are right!

**Comment by ludete** [comment]
OK.

**Pull request merged**